delivery_state/models/stock_picking.py

```diff
@@ -40,7 +41,34 @@ class StockPicking(models.Model):
     )
 
     def tracking_state_update(self):
-        for picking in self.filtered(lambda p: p.delivery_type):
+        """Call to the service provider API wich should have the method
+        definedin the manner. Can be triggered manually or by cron:
+        <my_provider>_tracking_state_update"""
+        for picking in self.filtered("delivery_type"):
             method = '%s_tracking_state_update' % picking.delivery_type
             if hasattr(picking.carrier_id, method):
                 getattr(picking.carrier_id, method)(picking)
+
+    @api.model
+    def _update_delivery_state(self):
+        """Automated action to query the delivery states to the carriers API.
+        every carrier should implement it 's own method. We split them by
+        delivery type so only those carries with the method update"""
+        pickings = self.search([
+            ("state", "=", "done"),
+            ("delivery_state", "not in",
+                ["customer_delivered", "canceled_shipment"]),
+            # These won't ever autoupdate, so we don't want to evaluate them
+            ("delivery_type", "not in", [False, "fixed", "base_one_rule"]),
+        ])
+        delivery_types = self.mapped("delivery_type")
```

Comment:
Instead of this, that requires several filters that are heavier than checking each time `hasattr`, I prefer a `read_group` grouped by `delivery_type`, and then process each group.

Reply:
The problem with that approach `delivery_type` is a related field, so a group_by can't be made with it :/  Anyway, the real time cost will be in the API queries, IMO. Also, we won't get much records (hundreds in an optimistic scenario, but dozens in the average one).

Reply:
It depends on the delivery volume you have, but OK, I see that is a non stored, so you can't group.